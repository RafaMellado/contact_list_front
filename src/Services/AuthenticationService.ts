import Cookies from "universal-cookie";
import { LoginRequestBody, LoginResponse } from "./Interfaces/Login";
import { SignUpRequestBody, SignUpResponse } from "./Interfaces/SignUp";
import LoginRepository from "./Repository/LoginRepository";
import SignUpRepository from "./Repository/SignUpRepository";

class AuthenticationService {
  async login(data: LoginRequestBody): Promise<void> {
    const response = await LoginRepository.login(data);

    if (response.token) {
      this.setAuthCookie(response);

      window.location.reload();
    }
  }

  async signUp(data: SignUpRequestBody): Promise<SignUpResponse> {
    return SignUpRepository.signUp(data);
  }

  logout(): void {
    const cookies = new Cookies();

    cookies.remove("contact-list-user");

    window.location.href = "/";
  }

  isLogged(): Boolean {
    const cookies = new Cookies();

    return !!cookies.get("contact-list-user");
  }

  setAuthCookie(data: LoginResponse): void {
    const cookies = new Cookies();

    cookies.set(
      "contact-list-user",
      { token: data.token, username: data.username },
      { expires: new Date(data.exp) }
    );
  }
}

export default new AuthenticationService();
