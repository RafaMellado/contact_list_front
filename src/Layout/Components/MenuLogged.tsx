import { Button } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import AuthenticationService from "../../Services/AuthenticationService";

export interface MenuLoggedProps {
  translations: string;
}

export function MenuLogged({ translations }: MenuLoggedProps) {
  const { t } = useTranslation();

  return (
    <>
      <Button
        className="me-3"
        variant="primary"
        data-testid="menu-logged-logout-btn"
        onClick={AuthenticationService.logout}
      >
        {t<string>(`${translations}.logout`)}
      </Button>
    </>
  );
}
