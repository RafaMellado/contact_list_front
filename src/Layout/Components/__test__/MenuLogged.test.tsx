import { MenuLogged } from "../MenuLogged";
import { act, fireEvent, render } from "@testing-library/react";
import type { MenuLoggedProps } from "../MenuLogged";
import AuthenticationService from "../../../Services/AuthenticationService";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock("../../../Services/AuthenticationService", () => ({
  logout: jest.fn(),
}));

describe(`<${MenuLogged.name} />`, () => {
  const menuLoggedProps: MenuLoggedProps = {
    translations: "contactHistory",
  };

  const factoryComponent = (props: MenuLoggedProps = menuLoggedProps) =>
    render(<MenuLogged {...props} />);

  test("renders correctly the component", () => {
    const component = factoryComponent();

    act(() => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  test("should execute logout function onClick", () => {
    const component = factoryComponent();

    act(() => {
      const element = component.getByTestId("menu-logged-logout-btn");

      fireEvent.click(element);
    });

    expect(AuthenticationService.logout).toHaveBeenCalled();
  });
});
