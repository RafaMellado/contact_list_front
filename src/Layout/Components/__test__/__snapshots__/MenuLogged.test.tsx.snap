// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuLogged /> renders correctly the component 1`] = `
<DocumentFragment>
  <button
    class="me-3 btn btn-primary"
    data-testid="menu-logged-logout-btn"
    type="button"
  >
    contactHistory.logout
  </button>
</DocumentFragment>
`;
